samples/rust/rust_semaphore.rs

```diff
@@ -111,31 +111,34 @@ impl FileOperations for FileState {
 }
 
 struct RustSemaphore {
-    _dev: Pin<Box<Registration<Arc<Semaphore>>>>,
+    _dev: Pin<Box<Registration<Ref<Semaphore>>>>,
 }
 
 impl KernelModule for RustSemaphore {
     fn init() -> Result<Self> {
         pr_info!("Rust semaphore sample (init)\n");
 
-        let sema = Arc::try_new(Semaphore {
-            // SAFETY: `condvar_init!` is called below.
-            changed: unsafe { CondVar::new() },
-
-            // SAFETY: `mutex_init!` is called below.
-            inner: unsafe {
-                Mutex::new(SemaphoreInner {
-                    count: 0,
-                    max_seen: 0,
-                })
+        let sema = Ref::try_new_and_init(
+            Semaphore {
+                // SAFETY: `condvar_init!` is called below.
+                changed: unsafe { CondVar::new() },
+
+                // SAFETY: `mutex_init!` is called below.
+                inner: unsafe {
+                    Mutex::new(SemaphoreInner {
+                        count: 0,
+                        max_seen: 0,
+                    })
+                },
             },
-        })?;
-
-        // SAFETY: `changed` is pinned behind `Arc`.
-        condvar_init!(Pin::new_unchecked(&sema.changed), "Semaphore::changed");
+            |sema| {
+                // SAFETY: `changed` is pinned when `sema` is.
+                condvar_init!(Pin::new_unchecked(&sema.changed), "Semaphore::changed");
```

Comment:
Shouldn't this be
```rust
condvar_init!(unsafe { Pin::new_unchecked(&sema.changed) }, "Semaphore::changed");
```
instead? Why is the missing `unsafe` not enforced?

That said, if this is indeed a problem, it should likely go in its own PR. So this comment shouldn't hold up the PR.

Reply:
It is because `init_with_lockdep!` wraps the expression inside `unsafe`.

Reply:
Nice catch, Sven! Uploaded #391 to fix this.